mava/networks/sable_network.py

```diff
@@ -0,0 +1,547 @@
+            )
+
+        # Compute the action logits
+        logit = self.head(x)
+
+        return logit, updated_hstates
+
+
+class SableNetwork(nn.Module):
+    """Sable network module."""
+
+    n_agents: int
+    action_dim: int
+    net_config: SableNetworkConfig
+    memory_config: DictConfig
+    action_space_type: str = "discrete"
+
+    def setup(self) -> None:
+        # Check if the action space type is valid
+        assert self.action_space_type in [
+            "discrete",
+            "continuous",
+        ], "Invalid action space type"
+
+        self.n_agents_per_chunk = self.n_agents
+        if self.memory_config.use_chunkwise:
+            if self.memory_config.type == "ff_sable":
+                self.memory_config.chunk_size = self.memory_config.agents_chunk_size
+                assert (
+                    self.n_agents % self.memory_config.chunk_size == 0
+                ), "Number of agents should be divisible by chunk size"
+                self.n_agents_per_chunk = self.memory_config.chunk_size
+            else:
+                self.memory_config.chunk_size = (
+                    self.memory_config.timestep_chunk_size * self.n_agents
+                )
+
+        # Create dummy decay scale factor for FF Sable
+        if self.memory_config.type == "ff_sable":
+            self.memory_config.decay_scaling_factor = 1.0
+        assert (
+            self.memory_config.decay_scaling_factor >= 0
+            and self.memory_config.decay_scaling_factor <= 1
+        ), "Decay scaling factor should be between 0 and 1"
+
+        # Decay kappa for each head
+        self.decay_kappas = 1 - jnp.exp(
+            jnp.linspace(jnp.log(1 / 32), jnp.log(1 / 512), self.net_config.n_head)
+        )
+        self.decay_kappas = self.decay_kappas * self.memory_config.decay_scaling_factor
+
+        self.encoder = Encoder(
+            self.net_config,
+            self.memory_config,
+            self.n_agents_per_chunk,
+        )
+        self.decoder = Decoder(
+            self.net_config,
+            self.memory_config,
+            self.n_agents_per_chunk,
+            self.action_dim,
+            self.action_space_type,
+        )
+
+        # Set the executor and trainer functions
+        (
+            self.train_encoder_fn,
+            self.train_decoder_fn,
+            self.execute_encoder_fn,
+            self.autoregressive_act,
+        ) = self.setup_executor_trainer_fn()
+
+    def __call__(
+        self,
+        obs_carry: Observation,
+        action: chex.Array,
+        hstates: HiddenStates,
+        dones: chex.Array,
+        rng_key: Optional[chex.PRNGKey] = None,
+    ) -> Tuple[chex.Array, chex.Array, chex.Array]:
+        """Training phase."""
+        # Get the observation, legal actions, and timestep id
+        obs, legal_actions, timestep_id = (
+            obs_carry.agents_view,
+            obs_carry.action_mask,
+            obs_carry.step_count,
+        )
```

Comment:
If you're going to do this, can't we just obs_carry explicitly?

Reply:
It was like that before and to be honest too much of params for a function so compressing them by getting full obs is the ideal